### PR TITLE
xtensa-build-zephyr: pass sof build version to rimage

### PR DESCRIPTION
**Pull request by dnikodem**
Rimage expects to receive fw_ver_build to prepare the manifest.

Signed-off-by: Damian Nikodem <damian.nikodem@intel.com>

### CODE REVIEW

**Comment by mwasko** [comment]
SOF CI

**Review by aborisovich — COMMENTED** [review]
This solution is generally bad - we have a version variables in CMake, we write them to header file and then we parse the header file to get variables back to CMake with regexes...
I do not like this at all and looking for better solutions to passing version to rimage.
I've tried using [configure_file](https://cmake.org/cmake/help/latest/command/configure_file.html) to create a sof_versions.h.in file replacing existing `sof_check_version_h` function but failed as Xtensa compilers (both commercial and from Zephyr SDK) do not have extensions to actually replace \#cmakedefine preprocessor (but that was just a refactor attempt). Maybe we can try something better with west sign command overload in sof?

**Review by marc-hb — CHANGES_REQUESTED** [review]
95% copy/paste/diverge of the function immediately above for no obvious reason.

**Comment by marc-hb** [comment]
> This solution is generally bad - we have a version variables in CMake, we write them to header file and then we parse the header file to get variables back to CMake with regexes...

Yes this is a bit of a hack but it's been working fine and hasn't caused any problem for months. We all have more urgent things to do but if you have time: the real problem is `west sign`. We shouldn't use `west sign`, we should sign directly from CMake just like XTOS still does.

I think we could not do sign directly from CMake in the past because SOF was not a real Zephyr application, SOF used to hijack `zephyr/samples/audio/sof` and `zephyr/samples/` cannot have a dependency on `rimage`; they're just samples But these days are over, since commit 00c407f8c9962af66 SOF is a real Zephyr application and having a CMake-time dependency on rimage would be fine.

> I've tried using [configure_file (https://cmake.org/cmake/help/latest/command/configure_file.html)...

I'm afraid a `configure_file` is generated at CMake configure-time, not at ninja build-time. So it would not be suitable.

**Comment by marc-hb** [comment]
> We shouldn't use west sign, we should sign directly from CMake just like XTOS still does.

Another problem caused by `west sign`: https://github.com/zephyrproject-rtos/zephyr/commit/9782b2cc0597fbe

**Comment by dnikodem** [comment]
> 

About "west sign -b" - according to west sign help, the west allows us to pass any parameter to the rimage tool directly after the "--" character (we are passing in the same way fw version with '-f' flag). 
According to the rimage code base, the rimage tool recognizes several command flags. One of them is '-b'.

**Review by lgirdwood — COMMENTED** [review]
@dnikodem I'm guessing you will validate on Windows ?
@marc-hb I assume CI will validate on Linux ?

**Comment by aborisovich** [comment]
> We shouldn't use west sign, we should sign directly from CMake just like XTOS still does.

Well that would be actually the easiest solution. Just use [add_custom_target](https://cmake.org/cmake/help/latest/command/add_custom_target.html) that DEPENDS on rimage_ep target (this target is defined in `src/arch/xtensa/CMakeLists.txt`: `ExternalProject_Add(rimage_ep`... Then you can trigger signing without using west at all with all needed cmake variables.

**Comment by lgirdwood** [comment]
> Then we could remove `west sign` command from `xtensa-build-zephyr.py` and forget about this problem for now.

I think only Intel and AMD sign FW currently. We should migrate to west though, and is the ```-b``` reuse between rimage and west is stopping us today ?

**Comment by aborisovich** [comment]
> > Then we could remove `west sign` command from `xtensa-build-zephyr.py` and forget about this problem for now.
> 
> I think only Intel and AMD sign FW currently. We should migrate to west though, and is the `-b` reuse between rimage and west stopping us today ?

No, nothing is stoping us. I'm not blocking anything just trying to provide better quality solutions because this one is not very good. As far as I spoke with @dnikodem this PR is not super urgent one or anything thus I'm trying to take a step in right direction.

**Review by aborisovich — APPROVED** [review]

File: scripts/xtensa-build-zephyr.py [hunk 1]
```diff
@@ -387,16 +388,17 @@ def west_update():
...
+	firmware file. When building multiple platforms from the same SOF
+	commit, all platforms share the same version. So for the 1st platform,
+	generate the version string from sof_version.h and later platforms will
+	reuse it.
 	"""
-	global sof_version
-	if sof_version:
-		return sof_version
+	global sof_fw_version
+	global sof_build_version
```
Comment (marc-hb) [inline scripts/xtensa-build-zephyr.py:399]: I'm fairly confident you don't need these `global` declarations. Otherwise looks good to me.

**Review by marc-hb — CHANGES_REQUESTED** [review]

**Comment by marc-hb** [comment]
> I think only Intel and AMD sign FW currently. 

This has been already solved with some CMake conditionals in the XTOS build, we could do more or less the same thing in the Zephyr application build.

>  @marc-hb what do you think?

I think we're on the same page for most or all topics. Good brainstorming for the longer term, in the mean time let's get this small and quick change merged.

**Comment by marc-hb** [comment]
> just realized the first global was already there. Still, please take the opportunity to fix this while you're changing this code (or prove me wrong).

I was wrong, apologies. the `global` keyword is needed (and confusing IMHO). The code in your previous commit bfc29a2db30b55e was correct, I should have approved it. Please restore it. Sorry for the hassle.

In your current commit 3371aad526b8c you made the variables local and removed the caching feature. I wasn't asking you to remove caching, I just thought the keyword was not needed to achieve caching. But the `global` keyword is required to implement caching, my bad.

I find `global` confusing because it is not required to _read_ a global variable, BUT it is required to _write_ to a global variable!

https://python-course.eu/python-tutorial/global-local-variables-namespaces.php

You can also (mis?)use `global` to _declare_ a global inside a function. I wrongly assumed it was it's only purpose.

PS: `global` is not required to use mutator methods: what a mess. I guess it achieves one of the purposes: discourage us from using globals.

**Review by kv2019i — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]

**Comment by marc-hb** [comment]
Unrelated suspend/resume failure in https://sof-ci.01.org/sofpr/PR6276/build1560/devicetest/
Unrelated license server failure in https://sof-ci.01.org/sofpr/PR6276/build1560/build/cnl_xcc_error.txt

**Comment by marc-hb** [comment]
SOFCI TEST

**Pull request merged**